### ISSUE TITLE
Consider using ghc-datasize for memory usage monitoring

### ISSUE THREAD

**Issue by lfairy** [haskell/hackage-server#403]
I recently found this package:

https://hackage.haskell.org/package/ghc-datasize

It has two advantages over the existing `MemSize` class:

* It's more accurate, since it looks at the internal heap structure directly

* It's more convenient, since we don't need to write an instance for every data type

**Comment by dcoutts** [comment]
Oh nice. Yes we should try that.

**Comment by dcoutts** [comment]
@lfairy unfortunately it looks like, as it's currently written, it's not suitable for large structures since it keeps a list of all the pointers it has visited.

https://hackage.haskell.org/package/ghc-datasize-0.1.2/docs/src/GHC-DataSize.html#recursiveSize

**Comment by lfairy** [comment]
Good catch! It appears that the purpose of that list is to avoid double-counting shared structures. But our `MemSize` machinery does not deal with sharing at all. So if we modify that code to remove the check, it should work with large structures without regressing from what we have now.

**Comment by ezyang** [comment]
Can someone tell me the motivation behind `MemSize`? As in, why is knowing the memory usage a sufficiently useful thing to have to merit this chunk of code?

**Comment by gbaz** [comment]
@dcoutts could fill in more details, but as I recall the main issue that dragged down deployment of this hackage rewrite was a very hard to manage memory footprint. All the memsize stuff was added in the course of needing to provide tooling to keep track of which features added what sort of footprint so that we could ensure that nothing was going to tank the memory footprint.

**Comment by ezyang** [comment]
I'm not sure about the wisdom of tying some auxiliary functionality to a pile of C/C-- code from ghc-heap-view, which would block us from moving to new versions of GHC until it was updated to handle any new closure types.

Maybe a better way of approximating the size is to GHC.Generics to get a generic representation of the type, and then read off the size based off of the number of fields in the generic representation. Even better, it records UNPACK, so we can even handle that case correctly. Maybe I'll prototype something up.
